network/network_utils_test.go

```diff
@@ -0,0 +1,168 @@
+}
+
+func (te *testSuite) startBoostrap(numNodes int, plugins ...network.PluginInterface) {
+	for i := 0; i < numNodes; i++ {
+		builder := network.NewBuilderWithOptions(te.builderOptions...)
+		builder.SetKeys(te.e.signature.RandomKeyPair())
+		builder.SetAddress(network.FormatAddress(te.e.networkType, "localhost", uint16(network.GetRandomUnusedPort())))
+
+		builder.AddPlugin(new(discovery.Plugin))
+		builder.AddPlugin(new(MailBoxPlugin))
+
+		for _, plugin := range plugins {
+			builder.AddPlugin(plugin)
+		}
+
+		node, err := builder.Build()
+		if err != nil {
+			te.t.Fatalf("Build() = expected no error, got %v", err)
+		}
+
+		go node.Listen()
+
+		if i == 0 {
+			te.bootstrapNode = node
+			node.BlockUntilListening()
+		} else {
+			te.nodes = append(te.nodes, node)
+		}
+	}
+
+	for _, node := range te.nodes {
+		node.Bootstrap(te.bootstrapNode.Address)
+	}
+
+	// wait for nodes to discover other peers
+	for _, node := range te.nodes {
+		pluginInt, ok := node.Plugin(discovery.PluginID)
+		if !ok {
+			te.t.Fatalf("Plugin() expected true, got false")
+		}
+
+		plugin := pluginInt.(*discovery.Plugin)
+		routes := plugin.Routes
+		peers := routes.GetPeers()
+
+		for len(peers) < numNodes-1 {
+			peers = routes.GetPeers()
+			time.Sleep(50 * time.Millisecond)
+		}
+	}
+}
+
+func (te *testSuite) tearDown() {
+	for _, node := range te.nodes {
+		node.Close()
+	}
+	te.bootstrapNode.Close()
+}
+
+func (te *testSuite) getMailbox(n *network.Network) *MailBoxPlugin {
+	if n == nil {
+		return nil
+	}
+	pluginInt, ok := n.Plugin(mailboxPluginID)
+	if !ok {
+		te.t.Errorf("Plugin(mailboxPluginID) expected true, got false")
+	}
+	return pluginInt.(*MailBoxPlugin)
+}
+
+func (te *testSuite) getPeers(n *network.Network) []peer.ID {
+	if n == nil {
+		return nil
+	}
+	pluginInt, ok := n.Plugin(discovery.PluginID)
+	if !ok {
+		te.t.Errorf("Plugin() expected true, got false")
+	}
+	plugin := pluginInt.(*discovery.Plugin)
+	routes := plugin.Routes
+	return routes.GetPeers()
+}
+
+// MailBoxPlugin buffers all messages into a mailbox for test validation.
+type MailBoxPlugin struct {
+	*network.Plugin
+	RecvMailbox chan *protobuf.TestMessage
+	SendMailbox chan *protobuf.TestMessage
+}
+
+// Startup creates a mailbox channel
+func (state *MailBoxPlugin) Startup(net *network.Network) {
+	state.RecvMailbox = make(chan *protobuf.TestMessage, 100)
+	state.SendMailbox = make(chan *protobuf.TestMessage, 100)
+}
+
+// Send puts a sent message into the SendMailbox channel
+func (state *MailBoxPlugin) Send(ctx *network.PluginContext) error {
+	switch msg := ctx.Message().(type) {
+	case *protobuf.TestMessage:
+		state.SendMailbox <- msg
+	}
+	return nil
+}
+
+// Receive puts a received message into the RecvMailbox channel
+func (state *MailBoxPlugin) Receive(ctx *network.PluginContext) error {
+	switch msg := ctx.Message().(type) {
+	case *protobuf.TestMessage:
+		state.RecvMailbox <- msg
+	}
+	return nil
+}
+
+func isIn(address string, ids ...peer.ID) bool {
```

Comment:
This one looks identical to the one in network/utils.go.

Reply:
I just didn't want to export this func for users. Maybe I should /shrug

Reply:
i think the test has access to it already though

Reply:
it's a `network_test` package, 'cause importing `discovery` causes a import cycle, which is forbidden in Go.
I will take a look, how we can address that.

Reply:
if it shares a lot of code, may be it doesn't need to be in the network_test package. Seems like these utility functions don't maintain any state nor too complicated, so the _test package may not need to be split out.